src/rust/engine/process_execution/src/remote.rs

```diff
@@ -70,50 +71,84 @@ impl CommandRunner {
     let execution_client2 = execution_client.clone();
     let operations_client = self.operations_client.clone();
     let store = self.store.clone();
-    let execute_request = make_execute_request(&req);
-    match execute_request {
-      Ok(execute_request) => {
-        let execute_request = Arc::new(execute_request);
-        future::ok(execute_request.clone())
-            .and_then(move |execute_request| {
-              map_grpc_result(execution_client.execute(&execute_request))
-            })
-            // TODO: Use some better looping-frequency strategy than a tight-loop.
-            // TODO: Add a timeout of some kind.
-            .and_then(move |operation| future::loop_fn(operation, move |operation| {
-              match (extract_execute_response(operation), store.clone()) {
-                (Ok(value), _) => {
-                  return future::ok(future::Loop::Break(value)).to_boxed() as BoxFuture<_, _>
+    let execute_request_result = make_execute_request(&req);
+
+    match execute_request_result {
+      Ok((command, execute_request)) => {
+        let execute_request_copy = execute_request.clone();
+        let store_copy = store.clone();
+        let store_copy2 = store.clone();
+        future::done(
```

Comment:
Extracting a method for the first 3 or 4 steps here would be good (basically: the "upload the command" portion).

Reply:
Done